client/state/initial-state.js

```diff
@@ -236,13 +237,19 @@ async function getInitialStoredState( initialReducer ) {
 
 	let initialStoredState = await getStateFromLocalStorage( initialReducer );
 	if ( ! initialStoredState ) {
-		return null;
+		initialStoredState = getEmptyState( initialReducer );
```

Comment:
This conditional return skipped going through the `storageKey`s in case the root state is not present in IndexedDB. We no longer want the early return: even when no stored state is present for the logged-in user, we want to load the logged-out `signup` slice.

The `getEmptyState` call here is suboptimal. We want the `getInitialStoredState` to behave as follows:
1. If there is no stored state that we want to return, then return `null`.
2. Return only the state subtree that was really present in IndexedDB. Don't add any "garbage", i.e., default values that were not stored.

This patched `getInitialStoredState` will return the default state of the reducer even if nothing was stored in IndexedDB at all. That's not necessary: the defaults are filled in elsewhere.

What about:
- removing the `initialStoredState` check completely
- in case it's null, let the `loadReducerState` function pass an empty object (`{}`) or `undefined` to the `initialReducer` instead of `null`. The reducers are not ready for `null`, as they either check for `undefined` or try to access properties of the input state.

That would keep `getInitialStoredState` sharply focused on one task and its unit tests wouldn't need to be changed.

Reply:
Thanks @jsnajdr, that's a much neater way to handle it. I was initially running into further errors passing in an empty object to `initialReducer`, but it turns out that was because of a line in `reducer-utils.js` that assumed `state[ key ]` could safely be accessed. I've replaced this with a lodash `get` call to safely return undefined if nothing's found, and now the tests are passing again: https://github.com/Automattic/wp-calypso/pull/34982/files#diff-279fd41426f3a4193fe0621d8d0ef23f, and I've reverted my changes to the existing tests (along with removing the `initialStoredState` check).

What do you think?